pkg/controller/controller.go

```diff
@@ -0,0 +1,388 @@
+		AddFunc:    ctrl.addPVC,
+		UpdateFunc: ctrl.updatePVC,
+		DeleteFunc: ctrl.deletePVC,
+	}, resyncPeriod)
+
+	return ctrl
+}
+
+func (ctrl *resizeController) addPVC(obj interface{}) {
+	objKey, err := getPVCKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
+}
+
+func (ctrl *resizeController) updatePVC(_, newObj interface{}) {
+	ctrl.addPVC(newObj)
+}
+
+func (ctrl *resizeController) deletePVC(obj interface{}) {
+	objKey, err := getPVCKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Forget(objKey)
+}
+
+func getPVCKey(obj interface{}) (string, error) {
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+	objKey, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Errorf("Failed to get key from object: %v", err)
+		return "", err
+	}
+	return objKey, nil
+}
+
+// Run starts the controller.
+func (ctrl *resizeController) Run(
+	workers int,
+	leaderElectionConfig *util.LeaderElectionConfig) {
+	run := func(ctx context.Context) {
+		defer ctrl.claimQueue.ShutDown()
+
+		klog.Infof("Starting external resizer %s", ctrl.name)
+		defer klog.Infof("Shutting down external resizer %s", ctrl.name)
+
+		stopCh := ctx.Done()
+
+		ctrl.informerFactory.Start(stopCh)
+		if !cache.WaitForCacheSync(stopCh, ctrl.pvSynced, ctrl.pvcSynced) {
+			klog.Errorf("Cannot sync pv/pvc caches")
+			return
+		}
+
+		for i := 0; i < workers; i++ {
+			go wait.Until(ctrl.syncPVCs, 0, stopCh)
+		}
+
+		<-stopCh
+	}
+
+	if leaderElectionConfig == nil {
+		// Leader election disabled.
+		run(context.TODO())
+	} else {
+		lock, err := util.NewLeaderLock(ctrl.kubeClient, ctrl.eventRecorder, leaderElectionConfig)
+		if err != nil {
+			klog.Fatalf("Error creating leader election lock: %v", err)
+		}
+		util.RunAsLeader(lock, leaderElectionConfig, run)
+	}
+}
+
+// syncPVCs is the main worker.
+func (ctrl *resizeController) syncPVCs() {
+	key, quit := ctrl.claimQueue.Get()
+	if quit {
+		return
+	}
+	defer ctrl.claimQueue.Done(key)
+
+	if err := ctrl.syncPVC(key.(string)); err != nil {
+		// Put PVC back to the queue so that we can retry later.
+		ctrl.claimQueue.AddRateLimited(key)
+	} else {
+		ctrl.claimQueue.Forget(key)
+	}
+}
+
+// syncPVC checks if a pvc requests resizing, and execute the resize operation if requested.
+func (ctrl *resizeController) syncPVC(key string) error {
+	klog.V(4).Infof("Started PVC processing %q", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("Split meta namespace key of pvc %s failed: %v", key, err)
+		return err
+	}
+
+	pvc, err := ctrl.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			klog.V(3).Infof("PVC %s/%s is deleted, no need to process it", namespace, name)
+			return nil
+		}
+		klog.Errorf("Get PVC %s/%s failed: %v", namespace, name, err)
+		return err
+	}
+
+	if !ctrl.pvcNeedResize(pvc) {
```

Comment:
I think just checking PVC alone here is not enough to prevent double processing of PVCs. For example - if user edits PVC and requests new size and while this resize is in-progress some other changes are done to same PVC. All such changes would trigger a resize.

Reply:
In some cases it may be unavoidable to double process same PVC but for most part I think we should avoid reconsidering a PVC which is already being resized. And that is why I think it may be worth comparing old vs new PVC size in update event.

Reply:
Sorry I missed the check in the update event handler when adjusting code structure.

Reply:
Added the check in update event handler and put the pvc into queue only if `newSize.Cmp(oldSize) > 0`, PTAL.